src/axom/quest/detail/MarchingCubesImpl.hpp

```diff
@@ -0,0 +1,998 @@
+
+#include "axom/config.hpp"
+
+// Implementation requires Conduit.
+#ifndef AXOM_USE_CONDUIT
+  #error "MarchingCubesImpl.hpp requires conduit"
+#endif
+#include "conduit_blueprint.hpp"
+
+#include "axom/core/execution/execution_space.hpp"
+#include "axom/slic/interface/slic_macros.hpp"
+#include "axom/quest/ArrayIndexer.hpp"
+#include "axom/quest/MeshViewUtil.hpp"
+#include "axom/quest/detail/MarchingCubesSingleDomain.hpp"
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/constants.hpp"
+#include "axom/mint/execution/internal/structured_exec.hpp"
+#include "axom/fmt.hpp"
+
+namespace axom
+{
+namespace quest
+{
+namespace detail
+{
+namespace marching_cubes
+{
+/*!
+  @brief Computations for MarchingCubesSingleDomain
+
+  Spatial dimension templating is here, to keep out of higher level
+  classes MarchCubes and MarchingCubesSingleDomain.
+
+  ExecSpace is the general execution space, like axom::SEQ_EXEC and
+  axom::CUDA_EXEC<256>.
+
+  See MarchingCubesImpl for the difference between that class and
+  MarchingCubesImpl.
+*/
+template <int DIM, typename ExecSpace, typename SequentialExecSpace>
+class MarchingCubesImpl : public MarchingCubesSingleDomain::ImplBase
+{
+public:
+  using Point = axom::primal::Point<double, DIM>;
+  using MIdx = axom::StackArray<axom::IndexType, DIM>;
+  using Indexer = axom::ArrayIndexer<axom::IndexType, DIM>;
+  using FacetIdType = int;
+  using LoopPolicy = typename execution_space<ExecSpace>::loop_policy;
+  using ReducePolicy = typename execution_space<ExecSpace>::reduce_policy;
+#if defined(AXOM_USE_RAJA)
+  // Intel oneAPI compiler segfaults with OpenMP RAJA scan
+  #ifdef __INTEL_LLVM_COMPILER
+  using ScanPolicy = typename axom::execution_space<axom::SEQ_EXEC>::loop_policy;
+  #else
+  using ScanPolicy = typename axom::execution_space<ExecSpace>::loop_policy;
+  #endif
+#endif
+  using SequentialLoopPolicy =
+    typename execution_space<SequentialExecSpace>::loop_policy;
+  static constexpr auto MemorySpace = execution_space<ExecSpace>::memory_space;
+
+  AXOM_HOST MarchingCubesImpl(int allocatorID,
+                              axom::Array<std::uint16_t>& caseIdsFlat,
+                              axom::Array<std::uint16_t>& crossingFlags,
+                              axom::Array<axom::IndexType>& scannedFlags,
+                              axom::Array<std::uint16_t>& facetIncrs)
+    : m_allocatorID(allocatorID)
+    , m_caseIds()
+    , m_caseIdsIndexer()
+    , m_caseIdsFlat(caseIdsFlat)
+    , m_crossingCases(0, 0, m_allocatorID)
+    , m_crossingFlags(crossingFlags)
+    , m_scannedFlags(scannedFlags)
+    , m_crossingParentIds(0, 0, m_allocatorID)
+    , m_facetIncrs(facetIncrs)  // (0, 0, m_allocatorID)
+    , m_firstFacetIds(0, 0, m_allocatorID)
+  {
+    SLIC_ASSERT(caseIdsFlat.getAllocatorID() == allocatorID);
+    SLIC_ASSERT(crossingFlags.getAllocatorID() == allocatorID);
+    SLIC_ASSERT(scannedFlags.getAllocatorID() == allocatorID);
```

Comment:
Gut check: Are these the only three arrays that should use the provided allocator?

Reply:
No.  `facetIncrs` should also be checked.  Thanks.